### ISSUE TITLE
report request - PDF: MVZ Ledger

### ISSUE THREAD

**Issue by dustymc** [ArctosDB/arctos#8365]
https://arctos.database.museum/Reports/reporter.cfm?action=edit&report_id=236

Possibly it's better to just tell me what you want instead of getting tangled up in the existing query, otherwise....

Possibly I can start by removing things that are in the query but not used, then ask whatever questions remain?? Other-otherwise....

I think hgeog is looking for``country: state_prov: county`` - yes?

elev is either number-units, number-number-units, or null?

lat_long - not sure, bunch of mixed data, help?

```
 case when strpos(flat.parts,'wing') > 0 then 'wing' else NULL end wing,
        case when strpos(flat.parts,'skeleton') > 0 then 'skeleton' else '' end skeleton,
        case when strpos(flat.parts,'skull') > 0 then 'skull' else '' end skull,
        case when strpos(flat.parts,'body') > 0 then 'mount' else '' end mount,
        case when strpos(flat.parts,'whole organism (ethanol)') > 0 then 'fluid-prep' else '' end fluid,
```

I think that's not used, can I drop it? Otherwise please let me know the intent.

tissues - not used, is something needed??

accn_number - what is needed?

mamm_meas
bird_meas

hu?

prep_number etc - what is needed, these will all always return NULL

**Comment by mkoo** [comment]
> https://arctos.database.museum/Reports/reporter.cfm?action=edit&report_id=236
> 
> Possibly it's better to just tell me what you want instead of getting tangled up in the existing query, otherwise....
> 
> Possibly I can start by removing things that are in the query but not used, then ask whatever questions remain?? Other-otherwise....

There's a lot of unnecessary junk for sure...

> 
> I think hgeog is looking for`country: state_prov: county` - yes?

correct (if county is null then skip to avoid a hanging ":" since we concatenate the fields here so could be handled better

> 
> elev is either number-units, number-number-units, or null?

I like number-number-units and null if null

> 
> lat_long - not sure, bunch of mixed data, help?

ugh, code is making me cringe, but yes dec_lat, dec_long, datum as WGS84 if exists

> 
> ```
>  case when strpos(flat.parts,'wing') > 0 then 'wing' else NULL end wing,
>         case when strpos(flat.parts,'skeleton') > 0 then 'skeleton' else '' end skeleton,
>         case when strpos(flat.parts,'skull') > 0 then 'skull' else '' end skull,
>         case when strpos(flat.parts,'body') > 0 then 'mount' else '' end mount,
>         case when strpos(flat.parts,'whole organism (ethanol)') > 0 then 'fluid-prep' else '' end fluid,
> ```
> 
> I think that's not used, can I drop it? Otherwise please let me know the intent.

I dont what the intent is/was- skip please

> 
> tissues - not used, is something needed??

not needed

> 
> accn_number - what is needed?

I really dont know what the intent of the SQL is with accession number-- we just need the accession number (we havent had a strong tradition of preceding zeros so just the value!)

> 
> mamm_meas bird_meas

I dont think the previous accessions had the measurements (and it's not in the current form) so skip pls (we have measurements on the box labels and this is simply a borrowed SQL from those reports)

> 
> hu?
> 
> prep_number etc - what is needed, these will all always return NULL

basically we need the original number which is either the collector or preparator number; rarely there are both so usually just one or the other
Thanks!

**Comment by dustymc** [comment]
Pre-update: 
[PDF_MVZ_Ledger.csv.zip](https://github.com/user-attachments/files/18283544/PDF_MVZ_Ledger.csv.zip)

I changed some stuff and cleaned up the SQL, let me know if that needs any further adjustments.

I'm not sure what's going on with the layout code, but I don't think it's going to result in valid HTML (and the server-side PDF-er will do near-random stuff if given invalid HTML). Let me know what the goal is if you need some help there.

**Comment by mkoo** [comment]
We're really close here to done but... this report needs some tweaks: https://arctos.database.museum/Reports/reporter.cfm?action=edit&report_id=236

@dusty help please!
- [ ] elevation = if min is same as max, print one of the values + units; if min not the same as max print ```*min* - *max*``` plus units (BTW the hyphen is not showing up currently so we'll test this)
- [ ] replace collection_cde "Mamm" with "Mammal" ( no need for the other collections)
- [ ] replace verbose datum with abbrecviation--> e.g. "WGS84" 
Other Ids is really messed up so I think I will have a complicated request next!

@birbjam help please!
- [ ] Have the header information of Accession number (flush left) and page number (flushed right) on the same line

**Comment by dustymc** [comment]
>elevation

done

>collection_cde "Mamm" with "Mammal"

done (?!)

>verbose datum with abbrecviation-

There's no datum in this so ???

**Comment by dustymc** [comment]
Actually maybe 

1. I possibly broke it, and
2. These manipulations are probably better done in cfml, which more people can change, rather than SQL


There were two elevation pulls, elev and elevation, I kept one, let me know if I picked the wrong one or can simplify for clean SQL and more complex layout per 2 above

**Comment by birbjam** [comment]
> [@birbjam](https://github.com/birbjam) help please!
> 
> * [x]  Have the header information of Accession number (flush left) and page number (flushed right) on the same line

@mkoo The report is protected, I can't edit and save it, but here's the CSS for your request. 

The following should help:

```css
.head {
  width: 100%;
  overflow: hidden;
}

.accn {
  float: left;
}

.page_right {
  float: right;
}

```

**Comment by mkoo** [comment]
@birbjam Thanks! For protected reports, just flip protected_template to ```false``` (dropdown) and save; then edit. It's just a little extra step to make sure you dont accidentally change a "finalized" report. If that is not working then let us know since it would be a bug

And yep those additions to CSS fixed it!

**Comment by mkoo** [comment]
also thanks for the SQL help

> elevation

Quick check-- looks ok but we need to test more

> Other ids

This needs more thinking on so let me formulate or put on a tech agenda

**Comment by mkoo** [comment]
Sorry it took a hot minute to get back to Other Ids in our ledger.

@dustymc  we need help with the SQL here:

![Image](https://github.com/user-attachments/assets/5ba7d89a-a6be-4a7f-9fa6-4b857bc8a1ea)

We want to edit the Other Ids so it does not show anything else BUT the identifier that matches with a name in the collector field. 
So in this example [MVZ:Bird:193438](https://arctos.database.museum/guid/MVZ:Bird:193438) above, if "Carla Cicero" appears in both collector and issued by (=agents match) then display "Carla Cicero (4317)" in Other Ids. We dont want to display any other other ids.

Let me know if I can explain better or a better idea than SQL function. The pro side for us is that it would take care of any collector AND preparator numbers; in fact I dont think we need to display that info since it doesnt necessarily help looking up info. (ie., we just need to know whose catalog to lookup an original number). In fact we're going to call it Original Id or something like that.
 thx!

**Comment by dustymc** [comment]
pretouch:

[PDF_MVZ_Ledger(1).csv.zip](https://github.com/user-attachments/files/18870670/PDF_MVZ_Ledger.1.csv.zip)

>Carla Cicero (4317)

Easily done (and later adjusted, if necessary) in CFML.

https://arctos.database.museum/guid/MVZ:Bird:171861

<img width="599" alt="Image" src="https://github.com/user-attachments/assets/e48f451f-acbd-4557-88ba-26df009a59c1" />

**Comment by mkoo** [comment]
huh, you did this as a cfml not sql ! Learn something new!  but it works only if the id_type is ```collector number``` ?

See https://arctos.database.museum/guid/MVZ:Bird:193633 or print out this as a ledger report:
https://arctos.database.museum/search.cfm?accn_trans_id=21129008
where  Orig No should be under Rauri Bowie (XXXX) and comes up empty

**Comment by dustymc** [comment]
OK, sorry, I didn't understand the scope.

Are you asking for a concatenation of all identifiers issued by any collector, regardless of role?? Eg if Phred had issued an identifier of arbitrary type on your example, you'd want something like ``Orig: Rauri C. K. Bowie (2474); Phred M. Benham (abc213)`` ??

Yes cfml is a full-blown language capable of very heavy lifting.

**Comment by dustymc** [comment]
>Yay cfml!

Well if you like it I'll stick with it!

Try now.

**Comment by mkoo** [comment]
As promised we have a new request, basically a second rule to show additional identifiers in a record.

We're calling it the Other Id rule: if there are other identifiers in a record where the Issuedby agent is an organization (eg another museum, NPS, university) then display: 
Other Id: agent's name-type of ```labels``` + value

This is a bit more involved since not all this info is in the current SQL -- puts the ' fun' in 'function'!

Here's a small test set with both personal catalog Orig numbers and Other Ids: https://arctos.database.museum/search.cfm?id_issuedby=UNLV&scientific_name=Ameiva

**Comment by dustymc** [comment]
Well there goes your cfml!

Pre-change:

[PDF_MVZ_Ledger.csv](https://github.com/user-attachments/files/18896659/PDF_MVZ_Ledger.csv)

There are two new functions:

 create mode 100644 function/label_print/getCollectorsIdentifiers.sql
 create mode 100644 function/label_print/getOrganizationIdentifiers.sql


The report is rebuilt to use them:

<img width="362" alt="Image" src="https://github.com/user-attachments/assets/8b9e35a2-9f27-4b59-8404-74f2b7a49c41" />

and a nice demo of dept of redundancy depting (https://github.com/ArctosDB/arctos/issues/8555#issuecomment-2672939323)

>Randy D. Jennings (RDJ 982)

**Comment by mkoo** [comment]
cool-- thanks! I can work with this
Also, yea we know how to redundancy more than once with duplication, too!
